### ISSUE TITLE
perl parser segv

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8292]
Migrated from [rt.perl.org#38299](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38299) (status was 'resolved')

Searchable as RT38299$

**Comment by p5pRT** [comment]
### From occ@cs.bath.ac.uk

#### Created by occ@noodle.cs.bath.ac.uk




There is a  bug in  the perl 5\.8\.\[5\-7\] \(at least\) parser which
manifests itself as a segfault while trying to parse syntactically
incorrect code\. 


The following code is the smallest example which i can make which
triggers the problem&#8203;:

sub foo \{
&nbsp;    my \($x\) = shift;
&nbsp;    sub \{\)\};
\}

However\, it seems that on some machines \(randomly?\) the attached longer
example is needed to trigger the bug \(the stack trace is identical\)\. 
\(the example file is a Parse&#8203;::Yapp  parser with bug and all\) 

stack trace&#8203;: 
\#0  Perl\_pad\_free \(po=5\) at pad\.c&#8203;:1171
\#1  0x08085e52 in Perl\_op\_clear \(o=0x81646b8\) at op\.c&#8203;:516
\#2  0x08085f88 in Perl\_op\_free \(o=0x81646b8\) at op\.c&#8203;:380
\#3  0x08085f4c in Perl\_op\_free \(o=0x81643e8\) at op\.c&#8203;:368
\#4  0x08085f4c in Perl\_op\_free \(o=0x81646d8\) at op\.c&#8203;:368
\#5  0x08085f4c in Perl\_op\_free \(o=0x8164700\) at op\.c&#8203;:368
\#6  0x08085f4c in Perl\_op\_free \(o=0x815f0b8\) at op\.c&#8203;:368
\#7  0x080910af in Perl\_newATTRSUB \(floor=109\, o=0x8151598\, proto=0x0\,
&nbsp;    attrs=0x0\, block=0x815f0b8\) at op\.c&#8203;:4387
\#8  0x0808513a in Perl\_yyparse \(\) at perly\.y&#8203;:364
\#9  0x080649e3 in S\_parse\_body \(env=0x0\, xsinit=0x805e670 \<xs\_init>\)
&nbsp;    at perl\.c&#8203;:1844
\#10 0x08065f21 in perl\_parse \(my\_perl=0x8149008\, xsinit=0x805e670
\<xs\_init>\,
&nbsp;    argc=2\, argv=0xbfcf31e4\, env=0x0\) at perl\.c&#8203;:1308
\#11 0x0805e61b in main \(argc=2\, argv=0xbfcf31e4\, env=0xbfcf31f0\)
&nbsp;    at perlmain\.c&#8203;:96


Note that this bug appears to be similar to\, if not a duplicate of 
\#34207 which was filed against perl 5\.8\.6 

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.7:

Configured by occ at Fri Jan 20 21:20:13 GMT 2006.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=linux, osvers=2.6.12-10-686, archname=i686-linux
    uname='linux noodle 2.6.12-10-686 #1 mon jan 16 17:58:04 utc 2006
i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe
-I/usr/local/include'
    ccversion='', gccversion='4.0.2 20050808 (prerelease) (Ubuntu
4.0.1-4ubuntu9)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.7:
    /usr/local/lib/perl5/5.8.7/i686-linux
    /usr/local/lib/perl5/5.8.7
    /usr/local/lib/perl5/site_perl/5.8.7/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.7
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.7:
    HOME=/home/occ/
    LANG=en_GB
    LANGUAGE=en_GB
    LD_LIBRARY_PATH=/usr/lib:/usr/local/lib
    LOGDIR (unset)

PATH=/opt/jdk1.5.0_01//bin:/opt/apache-ant-1.6.2/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11:/usr/games:/home/occ/linux/bin/
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From occ@cs.bath.ac.uk

[bug.pm](https://rt-archive.perl.org/perl5/Ticket/Attachment/128631/86798/bug.pm)

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Jan 21\, 2006 at 07&#8203;:11&#8203;:12AM \-0800\, Owen Cliffe wrote&#8203;:
> There is a  bug in  the perl 5\.8\.\[5\-7\] \(at least\) parser which
> manifests itself as a segfault while trying to parse syntactically
> incorrect code\. 
> 
> 
> The following code is the smallest example which i can make which
> triggers the problem&#8203;:
> 
> sub foo \{
>     my \($x\) = shift;
>     sub \{\)\};
> \}

Thanks for the report\, sorry for the lateness of a response\.

This bug has been fixed in development branch of perl&#8203;: what will
eventually become 5\.10\.0\. There are no plans to backport it to the 5\.8\.x
branch\.

\-\- 
The optimist believes that he lives in the best of all possible worlds\.
As does the pessimist\.
